### PR TITLE
Allow Blocks for Custom Product taxonomies

### LINKED ISSUE #31607: Undefined $default_file for product taxonomy templates
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

This issue is related with #30013

In `WC_Template_Loader:get_template_loader_default_file` the goal is to fallback to PHP template version in case the expected block template is not found.

It works as expected if most of the case, but for product taxonomy there is a bug.

If you crate a product taxonomy like `taxonomy_product_brands` the code will enter in the if condition `is_product_taxonomy()` 

But since  `is_tax( 'product_cat' ) || is_tax( 'product_tag' )` will return false...  then it will never check if the block.`taxonomy-product_brand.html` is present. 

As next step it will check if the system has `archive-product` template. 

- In case it isn't it will load `archive-product.php` which produces a deprecation WARNING (doesn't matter if you have the taxonomy bock or not).
- In case it is present, it will go out the condition and `$default_file` will be undefined creating a NOTICE message in PHP.

Screenshot 2022-01-11 at 11 22 36

### Expected behavior

In case you load `taxonomy_product_brands` (or any other taxonomy block template) the function should check it there is a block for that template `$default_file` should be '' instead of undefined

In case  `archive-product` is present, `$default_file` should be '' instead of undefined

### Actual behavior

- Function doesn't check if there is a block for product taxonomies other than `product_cat` or `product_tag`
- Function returns undefined `$default_file` if archive-product is present.

### Steps to reproduce

<details>
<summary>
taxonomy-product_brands.html content
</summary>

```
<!-- wp:template-part {"slug":"header"} /-->
<!-- wp:group {"layout":{"inherit":true}} -->
<div class="wp-block-group"><!-- wp:woocommerce/legacy-template {"template":"archive-product"} /--></div>
<!-- /wp:group -->
<!-- wp:template-part {"slug":"footer"} /-->
```

</details>

**Manual test with your own Product Taxonomy** 

1. Install WC 6.1
2. Create a custom product taxonomy. For example taxonomy_product_brands
3. Load a block theme like TT2
4. Load the file `taxonomy-product_brands.html` provided at the top in the theme `templates` folder
5. Navigate to the taxonomy archive in the frontend. 
6. You will get a notice message if archive-product.html is present
7. You will get a warning if archive-product.html is not present

**Alternative: Test with a plugin like WooCommerce Product Brands avoiding creating the Taxonomy manually**

1. Install WC 6.1
2. Install WooCommerce Product Brands and create a brand in Product -> Brands
3. Load a block theme like TT2
4. Load the file `taxonomy-product_brands.html` provided at the top in the theme `templates` folder
5. Navigate to the taxonomy archive in the frontend. 
6. You will get a notice message if archive-product.html is present
7. You will get a warning if archive-product.html is not present

### WordPress Environment ###

WordPress address (URL): https://60e8-213-94-23-19.ngrok.io
Site address (URL): https://60e8-213-94-23-19.ngrok.io
WC Version: 6.1.0
REST API Version: ✔ 6.1.0
WC Blocks Version: ✔ 6.5.1
Action Scheduler Version: ✔ 3.4.0
WC Admin Version: ✔ 3.0.1
Log Directory Writable: ✔
WP Version: ❌ 5.9-RC1-52542 - There is a newer version of WordPress available (6.0-alpha-52549)
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: –
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.21.4
PHP Version: 7.4.25
PHP Post Max Size: 0 B
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.68.0
OpenSSL/1.1.1f

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.3.32-MariaDB-1:10.3.32+maria~focal
Max Upload Size: 0 B
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 5.9.0
WC Database Prefix: wp_
Total Database Size: 8.59MB
Database Data Size: 5.50MB
Database Index Size: 3.09MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.17MB + Index: 0.14MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.11MB + Index: 0.09MB + Engine InnoDB
wp_automatewoo_abandoned_carts: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_automatewoo_customers: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_automatewoo_customer_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_automatewoo_events: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_automatewoo_guests: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_automatewoo_guest_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_automatewoo_logs: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_automatewoo_log_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_automatewoo_queue: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_automatewoo_queue_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.05MB + Index: 0.09MB + Engine InnoDB
wp_followup_coupons: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_followup_coupon_logs: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_followup_customers: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_followup_customer_carts: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_followup_customer_notes: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_followup_customer_orders: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_followup_email_excludes: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_followup_email_logs: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_followup_email_orders: Data: 0.05MB + Index: 0.09MB + Engine InnoDB
wp_followup_email_tracking: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_followup_followup_history: Data: 0.09MB + Index: 0.05MB + Engine InnoDB
wp_followup_order_categories: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_followup_order_items: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_followup_subscribers: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_followup_subscribers_to_lists: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_followup_subscriber_lists: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_gla_budget_recommendations: Data: 0.22MB + Index: 0.14MB + Engine InnoDB
wp_gla_merchant_issues: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_gla_shipping_rates: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_gla_shipping_times: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 3.05MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.22MB + Index: 0.20MB + Engine InnoDB
wp_posts: Data: 0.09MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.05MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wpmailsmtp_debug_events: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wpmailsmtp_tasks_meta: Data: 0.02MB + Index: 0.00MB + Engine InnoDB

### Post Type Counts ###

attachment: 23
follow_up_email: 8
page: 8
post: 4
product: 19
product_variation: 7
revision: 10
shop_order: 16
shop_order_refund: 33
wp_global_styles: 1
wp_template: 1
wp-template: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (8) ###

Google Listings and Ads: by WooCommerce – 1.9.0
WCS Variables: by WooCommerce – 0.1
WooCommerce Brands: by WooCommerce – 1.6.25
WooCommerce Stripe Gateway: by WooCommerce – 5.9.0
WooCommerce Google Analytics Integration: by WooCommerce – 1.5.6
WooCommerce: by Automattic – 6.1.0-rc.2
WordPress Beta Tester: by Peter Westwood
Andy Fragen – 3.1.4

WP Mail SMTP: by WPForms – 3.2.1

### Inactive Plugins (6) ###

Akismet Anti-Spam: by Automattic – 4.2.1
AutomateWoo: by WooCommerce – 5.5.7
Follow-Up Emails: by WooCommerce – 4.9.16
Pinterest for WooCommerce: by WooCommerce – 1.0.1
WooCommerce Admin: by WooCommerce – 3.0.1
WooCommerce Beta Tester: by WooCommerce – 2.0.5

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –

### WC Pages ###

Shop base: #5 - /shop/
Cart: #6 - /cart/
Checkout: #7 - /checkout/
My account: #8 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Two
Version: 1.0
Author URL: https://wordpress.org/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ❌ Not declared

### Templates ###

Overrides: –

### Action Scheduler ###

Complete: 387
Oldest: 2021-12-29 16:00:39 +0000
Newest: 2022-01-11 10:32:09 +0000

Pending: 3
Oldest: 2022-01-11 10:49:14 +0000
Newest: 2022-01-12 03:00:00 +0000

### Status report information ###

Generated at: 2022-01-11 10:39:22 +00:00

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by puntope**
### All Submissions:

* [ x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #31607

We added here #30013 some logic for checking Block templates. It works well in most of the cases. However, it doesn't work for custom product taxonomies. 

This PR fixes that. 

### How to test the changes in this Pull Request:

<details>
<summary>
taxonomy-product_brands.html content
</summary>

```
<!-- wp:template-part {"slug":"header"} /-->
<!-- wp:group {"layout":{"inherit":true}} -->
<div class="wp-block-group"><!-- wp:woocommerce/legacy-template {"template":"archive-product"} /--></div>
<!-- /wp:group -->
<!-- wp:template-part {"slug":"footer"} /-->
```
</details>

**Manual test with your own Product Taxonomy** 

1. Clone this PR
2. Create a custom product taxonomy. For example `taxonomy_product_brands` (see alternative testing method for preventing creating manually the taxonomy)
3. Load a block theme FSE like TT2
4. Load the file `taxonomy-product_brands.html` provided at the top in the theme `templates` folder
5. Navigate to the taxonomy archive in the frontend. 
6. You won't get a notice neither warnings, and the template will load. 
7. Check how other templates like Product Category, Product Tag and Shop Archive Product load as always.

**Alternative: Test with a plugin like WooCommerce Product Brands avoiding creating the Taxonomy manually**

1. Clone this PR
2. Install [WooCommerce Product Brands](https://woocommerce.com/products/product-brands-for-woocommerce/) and create a brand navigating to Product -> Brands
3. Load a block theme with FSE like TT2
4. Load the file `taxonomy-product_brands.html` provided at the top in the theme `templates` folder
5. Navigate to the taxonomy archive in the frontend. 
6. You won't get a notice neither warnings, and the template will load. 
7. Check how other templates like Product Category, Product Tag and Shop Archive Product load as always.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Fix - Support custom Product Taxonomies in block template loader

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Review by sunyatasattva — COMMENTED** [review]
In general, this LGTM, but I am unsure about the original intention of this piece of code, so maybe calling someone involved with the original creation or doing some digging on PRs could be good.

My doubt here is the following:

The previous code had a fallback to `'taxonomy-' . $object->taxonomy . '.php'` in case we did not have a block template for that specific taxonomy. But only considered `product_cat` and `product_tag` as valid taxonomies.

Now that we have expanded support for custom taxonomies, wouldn't that make sense to also just generally falling back to the appropriate `'taxonomy-' . $object->taxonomy . '.php'` regardless?

In a nutshell, I mean, shouldn't we remove the guard clause around that altogether? As far as I understand this code, that was the original intention.

```
Is this a specific taxonomy?
  YES → Do we have the right block template?
    YES → Cool, we don't need a fallback file.
    NO → Fallback to the taxonomy php archive file
  NO → Do we have the block template for the `archive-product`?
    YES → Cool, let's use that.
    NO → Let's use `archive-product.php`
```

**Comment by puntope** [comment]
> The previous code had a fallback to `'taxonomy-' . $object->taxonomy . '.php'` in case we did not have a block template for that specific taxonomy. But only considered `product_cat` and `product_tag` as valid taxonomies.

As far as I understood this part of code, this is for a safe fallback to the PHP template. Correct me if I'm wrong but `taxonomy-product-cat.php` and `taxonomy-product-tag.php` together with `archive-product.php` are always present in WooCommerce. Whereas the custom taxonomy PHP file could be not present, hence it could fallback to an undefined template creating side effects.

One option could be to check as well the PHP templates  and do some logic regarding that, but IMHO that's out scope this PR which just tries to solve the undefined `$default_file` error and adds ability to load custom taxonomy-product html templates without creating a forced fallback always to Archive product.

**Review by vedanshujain — COMMENTED** [review]
LGTM! Can you add some unit tests as well, this has grown to be a bit complicated, unit tests will help us make maintain and make changes to this code later.

**Comment by puntope** [comment]
> LGTM! Can you add some unit tests as well, this has grown to be a bit complicated, unit tests will help us make maintain and make changes to this code later.

Howdy @vedanshujain As requested... I added relevant Unit tests for `WC_Template_Loader::template_loader` which covers as well `WC_Template_Loader::get_template_loader_default_file`

File: plugins/woocommerce/includes/class-wc-template-loader.php [hunk 1]
```diff
@@ -159,6 +159,7 @@ private static function has_block_template( $template_name ) {
 	 * @since  3.0.0
 	 * @since  5.5.0 If a block template with the same name exists, return an
 	 * empty string.
+	 * @since  6.1.0 It checks custom product taxonomies
```
Comment (vedanshujain) [inline plugins/woocommerce/includes/class-wc-template-loader.php]:
```suggestion
	 * @since  6.3.0 It checks custom product taxonomies
```

This should be 6.3.0 now I think.

**Review by vedanshujain — DISMISSED** [review]
LGTM! Thanks for adding unit tests, there is a minor version change, otherwise feel free to merge.

**Review by vedanshujain — APPROVED** [review]
